lib/pavilion/output.py

```diff
@@ -72,6 +73,48 @@
     'UNDERLINE': UNDERLINE,
 }
 
+DEFAULT_FORMAT = '{0}'
+
+
+def _num_digits(n: int) -> int: # pylint: disable=invalid-name
+    """Only attempt to count number digits for ints, since
+    precision error makes it nonsensical for floats."""
+    if n < 0:
+        n = abs(n)
+    elif n == 0:
+        return 1
+    return floor(log10(n)) + 1
+
+
+def format_numeric(value: Union[int, float], digits: Optional[int]) -> str:
+    if digits is None:
+        return str(value)
+
+    sci_fmt = '{' + f':.{digits-1}e' + '}'
+
+    if isinstance(value, int):
+        if _num_digits(value) > digits:
+            res = sci_fmt.format(value)
+        else:
+            res = str(value)
+    elif isinstance(value, float):
+        if abs(value) < 10**-(digits - 1):
+            res = sci_fmt.format(value)
+        else:
+            res = str(round(value, digits - 1))
+    else:
+        raise ValueError(f'Cannot format value {value}. Expected one of (int, float),' +
+            f' but received {type(value)}.')
+
+    # Remove leading zeros in exponent, as well as + sign, if present
+    pos_regex = 'e\+0*'
+    neg_regex = 'e-0*'
+
+    res = re.sub(pos_regex, "e", res)
+    res = re.sub(neg_regex, "e-", res)
+
```

Comment:
I had all kinds of complaints, realized why you were doing all the things I wanted to complain about, and ran out of complaints.